Blog post: how we integrate with Git (#843)
fluxcd/website: content/en/blog/2022-03-25-putting-the-git-into-gitops/index.md

---
author: dholbach
date: 2022-03-25 09:30:00+00:00
title: 'Flux puts the Git into GitOps'
description: Flux integrates very tightly with all relevant APIs and SDKs. For us to provide the best possible Git support to bring you GitOps, shelling out to Git is not an option. Find out why in this blog post.
url: /blog/2022/03/flux-puts-the-git-into-gitops/
---

Ever since the rewrite of Flux as a set of focused controllers, it has
become clearer what each of its functions and capabilities are. The
aptly named controllers carry in their name what they are responsible
for and which data or tooling they interact with, so that is, e.g.
`source`, `kustomize`, `image-automation`, `notification`, `helm`,
etc.

![Overview Flux controllers](/img/diagrams/gitops-toolkit.png)

If you wanted to string a proof-of-concept for a GitOps tool together, a
naïve solution could be to just shell out to various tools like `curl`,
`git`, `kubectl` and `helm`. While that might feel intuitive at first (since

> it so closely resembles ones manual workflow behind the keyboard) and

Reviewer: ```suggestion
it so closely resembles one's manual workflow behind the keyboard) and
```
maybe?

security. Of course we want to pass this all down to our users: more
efficient download makes a huge difference, support for Git Submodules
enables new use-cases, support for more GPG verification or new SSH key
formats adds additional security and when new features are rolled out in
Git providers, we need to support them in Flux too.

Integrating these changes into Flux unfortunately isn't as easy as it
sounds. One part of the dependency chain for
[git2go](https://github.com/libgit2/git2go) goes:

```mermaid
flowchart LR
libgit2 --> libssh2 --> OpenSSL
```

So that's [libgit2](https://libgit2.org/),
[libssh2](https://libssh2.org) (to enable the SSH transport) and
[OpenSSL](https://www.openssl.org/). As Linux vendors often take a very
conservative approach to bringing new software releases to stable
releases, we were unfortunately pushed to
[building these dependencies ourselves](https://github.com/fluxcd/golang-with-libgit2#rationale).
In addition to that, these libraries have quite a few configuration
options that can only be set at build time and unfortunately the
`openssl`/`libssh2` packages of different Linux distributions act in
[slightly different

> ways](https://github.com/fluxcd/golang-with-libgit2/blob/main/hack/Makefile#L63-L69).

Reviewer: A versioned link would be better here to avoid pointing to unrelated section in the future.
```suggestion
ways](https://github.com/fluxcd/golang-with-libgit2/blob/libgit2-1.3.0/hack/Makefile#L63-L69).
```

similar issues, so to avoid delaying a fix or forking the dependency, we
decided to add experimental support for Go managed transports, which means
we can enforce that network operations won't take more than a given
amount of time to complete, but without requiring any changes upstream.

This is [part of Flux 0.28](https://github.com/fluxcd/flux2/releases/tag/v0.28.0)
and can be enabled by adding an environment `EXPERIMENTAL_GIT_TRANSPORT=true`
in both `source` and `image-automation` controllers.

This will give us more control over the transport with Go native
transport using the `libgit2` smart transport support. Read the
[source-controller
changelog](https://github.com/fluxcd/source-controller/blob/main/CHANGELOG.md#experimental-managed-transport-for-libgit2-git-implementation)
for more information.

If you want to enable this automatically, just add the following to
your `kustomization.yaml`:

```yaml
patches:
- patch: |
    - op: add
      path: /spec/template/spec/containers/0/env/0
      value:
        name: EXPERIMENTAL_GIT_TRANSPORT
        value: "true"
  target:
    kind: Deployment
    name: "(source-controller|image-automation-controller)"
```

> We are also adding support for sha256 hashes: Git supports it since 2.29
> but all major Git service provides such as GitLab and GitHub still don't.
> We are watching this to be able to adopt support as the industry moves
> on from sha1.

Reviewer 2: ```suggestion
Support for sha256 hashes: Git CLI supports it since 2.29, however all major Git
service providers, such as GitLab and GitHub, are yet to make some progress on 
this front.
Upstream, libgit2 is starting to pave the way for that support on [v1.4.0]https://github.com/libgit2/libgit2/releases/tag/v1.4.0),
we are watching this space to be able to support Flux users as the industry moves on
from SHA1.
```
